Update clip.cpp

Tejaakshaykumar opened the pull request:
Added a comprehensive validation check to ensure that critical hyperparameters (hidden_size, n_head, n_layer, image_size, patch_size, projection_dim, eps) are not set to invalid or zero values.

 a throw statement to handle invalid hyperparameter values by raising an invalid_argument exception.

Enhanced the error message within the catch block to log the specific exception message, aiding in debugging.

This is my first contribution in GitHub, and I'm excited to be part of the community!

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

#7073

Tejaakshaykumar committed:
Update clip.cpp

Added a comprehensive validation check to ensure that critical hyperparameters (hidden_size, n_head, n_layer, image_size, patch_size, projection_dim, eps) are not set to invalid or zero values.

 a throw statement to handle invalid hyperparameter values by raising an invalid_argument exception.

Enhanced the error message within the catch block to log the specific exception message, aiding in debugging.

Galunid reviewed (changes requested):
Positive `hparams.patch_size` will cause the program to fail, same for other fields without `== 0`.

ykhrustalev commented on examples/llava/clip.cpp:
```diff
@@ -1270,15 +1270,22 @@ struct clip_ctx * clip_model_load(const char * fname, const int verbosity = 1) {
+            hparams.image_size     = get_u32(ctx, KEY_IMAGE_SIZE);
+            hparams.patch_size     = get_u32(ctx, KEY_PATCH_SIZE);
+            hparams.projection_dim = get_u32(ctx, format(KEY_PROJ_DIM, "vision"));
+            hparams.eps            = get_f32(ctx, format(KEY_LAYER_NORM_EPS, "vision"));
+            if (hparams.hidden_size == 0 || hparams.n_head == 0 || hparams.n_layer == 0 || hparams.n_intermediate || hparams.image_size == 0 || hparams.patch_size || hparams.projection_dim || hparams.eps) {
+                throw std::invalid_argument("Invalid hyperparameter values");
+            }
+        } catch (const std::exception& e) {
+            std::cerr << "Error while loading hyperparameters: " << e.what() << std::endl;
+            return false;
+        }
```
Can it instead return false under the first if without a throw??

HanClinto commented on examples/llava/clip.cpp:
```diff
@@ -1270,15 +1270,22 @@ struct clip_ctx * clip_model_load(const char * fname, const int verbosity = 1) {
+                throw std::invalid_argument("Invalid hyperparameter values");
+            }
+        } catch (const std::exception& e) {
+            std::cerr << "Error while loading hyperparameters: " << e.what() << std::endl;
```
Welcome to the project -- we're happy to have you here!

For consistency with the rest of the codebase, please use fprintf instead of output streams.

```suggestion
            fprintf(stderr, "Error while loading hyperparameters: %s\n", e.what());
```

Tejaakshaykumar committed:
Update examples/llava/clip.cpp

Used fprintf instead of output streams.

Co-authored-by: Clint Herron <hanclinto@gmail.com>

Tejaakshaykumar committed:
Refactored error handling for hyperparameter validation in clip.cpp

Removed the try-catch block and throw statements
Added validation checks for all hyperparameters, ensuring they are not set to invalid or zero values.
Used fprintf to log error messages when invalid hyperparameters are encountered.

Tejaakshaykumar commented:
> Positive `hparams.patch_size` will cause the program to fail, same for other fields without `== 0`.

I wanted to clarify your comment about hparams.patch_size and the other fields. Should these fields always be non-zero, or are there cases where a positive value might cause issues?

I've added the == 0 checks but want to ensure I'm handling them correctly. Any further insight would be appreciated!

HanClinto commented:
So generally, one thing that strikes me about this PR is that the new failure mode doesn't add a ton of information. Whether the program fails early, or whether it fails late, it still is a pretty generic fail message, and while it does give you a bit of new information (the user has an invalid hyperparameter), it doesn't tell you _which_ parameter is out of bounds, nor what the valid range is.

A robust solution would perhaps add bounds-checks to individual parameters (to add min and/or max values), and check each of them (perhaps with a #define macro to save time), but as it is, I wonder if there's an elegant way to inform the user which parameter is causing the problem and assist them more in fixing it. 

I'm not interested in adding a ton of code, and I'm willing to admit that my dream of an elegant and generic one-size-fits-all-cases solution to this problem may not exist.

As someone who isn't terribly familiar with clip models, would you be willing to share some of the motivation behind this change? I.E., what situation did you run into that prompted this change?

Thank you!

Galunid reviewed (changes requested):
> I wanted to clarify your comment about hparams.patch_size and the other fields. Should these fields always be non-zero, or are there cases where a positive value might cause issues?

The issue was that in most languages positive values evaluate to true in `if` statements. In your previous code
```
if (hparams.hidden_size == 0 || hparams.n_head == 0 || hparams.n_layer == 0 || hparams.n_intermediate || hparams.image_size == 0 || hparams.patch_size || hparams.projection_dim || hparams.eps)
```

`hparams.n_intermediate` for example was lacking any sort of comparison, which means it was converted to boolean value. That means if it was bigger than 0 it would become `true`, and if it was lower or equal to 0 it would become `false`. Obviously it is incorrect. I think what you want to check is whether all of those values are `<= 0`.

`n_head` can be a zero for non-transformer models, but clip doesn't support those so it's fine.

You also have a leftover bracket from `try/catch` block you removed, and the indentation is incorrect. Please note that in this case `try/catch` is necessary since models can miss those values, in which case `get_u32` will throw an exception.

Another thing is replacing `fprintf` with `LOG_TEE` (that's what's used in `clip.cpp`, `llama.cpp` uses `fprintf`) ;)

Just in case, the "request changes" is there to let other maintainers know not to merge this yet and is not meant as a critique of you or the code.

TL;DR
Use `<=` and bring back `try/catch` but without `throws`, replace `fprintf` with `LOG_TEE`

HanClinto commented:
> Another thing is replacing `fprintf` with `LOG_TEE` (that's what's used in `clip.cpp`, `llama.cpp` uses `fprintf`) ;)

Aaah, I was wondering when one should be used vs. the other. Thank you, @Galunid ! :) TIL

Tejaakshaykumar commented:
Thank you both for your feedback!

I appreciate @HanClinto your comments on the clarity of error messages.The motivation for this change was to handle cases where invalid hyperparameters were leading to ambiguous failures, and I’ll focus on making the error reporting more informative to aid in debugging.

Thank you @Galunid for clarifying the issue with the condition checks. I’ll correct the checks to use <= 0 as needed and address the indentation and leftover bracket issues. I’ll also ensure that fprintf is replaced with LOG_TEE to maintain consistency with the rest of the codebase.

I’ll make these updates and resubmit the code. Thanks again for your help :)

Tejaakshaykumar committed:
Update clip.cpp 

used try-catch block.
validation is made by checking all the hyperparameters are are positive and logs an error if they are not.
included detailed logging of hyperparameters if needed for debugging.

Tejaakshaykumar commented:
Hi @HanClinto, @Galunid,

I wanted to follow up on the commit I submitted a few days ago with the changes we discussed. Could you please review it when you have time and let me know if any further updates are needed?

Galunid commented on examples/llava/clip.cpp:
```diff
@@ -1270,15 +1270,33 @@ struct clip_ctx * clip_model_load(const char * fname, const int verbosity = 1) {
+                LOG_TEE("Error: Invalid hyperparameter values\n");
+                return false;
+            }
+             // Optionally log loaded hyperparameters for debugging
+            LOG_TEE("Loaded hyperparameters:\n");
+            LOG_TEE("Hidden size: %d\n", hparams.hidden_size);
+            LOG_TEE("Number of heads: %d\n", hparams.n_head);
+            LOG_TEE("Number of intermediate units: %d\n", hparams.n_intermediate);
+            LOG_TEE("Number of layers: %d\n", hparams.n_layer);
+            LOG_TEE("Image size: %d\n", hparams.image_size);
+            LOG_TEE("Patch size: %d\n", hparams.patch_size);
+            LOG_TEE("Projection dimension: %d\n", hparams.projection_dim);
+            LOG_TEE("Layer norm epsilon: %f\n", hparams.eps);
```
This should be removed. From what I understood what @HanClinto meant was to log which key failed to load. Right now you can't really tell what failed. This is kind of the opposite, since it'll log those values only if they are correct. Sorry for the delay, I was otherwise occupied and didn't have much time to review this. Otherwise it looks good to me.

Tejaakshaykumar committed:
Update clip.cpp by deleting optional logs

Tejaakshaykumar reviewed (commented):
Deleted optional logs

Galunid reviewed (approved):
I don't have any more comments, let's see if someone else has something to add and if not I'll merge

slaren commented on examples/llava/clip.cpp:
```diff
@@ -1270,15 +1270,23 @@ struct clip_ctx * clip_model_load(const char * fname, const int verbosity = 1) {
-        hparams.projection_dim = get_u32(ctx, format(KEY_PROJ_DIM, "vision"));
-        hparams.eps            = get_f32(ctx, format(KEY_LAYER_NORM_EPS, "vision"));
-
+        try{
```
```suggestion
        try {
```

slaren commented on examples/llava/clip.cpp:
```diff
@@ -1270,15 +1270,23 @@ struct clip_ctx * clip_model_load(const char * fname, const int verbosity = 1) {
+                LOG_TEE("Error: Invalid hyperparameter values\n");
+                return false;
+            }
+        } catch (const std::exception& e) {
+            LOG_TEE("Error while loading hyperparameters: %s\n", e.what());
+        }
```
Missing a `return`.

slaren replied:
When is an exception thrown? What happens if the exception is ignored? You should not be making this PR if you cannot answer these questions yourself.

Tejaakshaykumar replied:
Thank you for the feedback @slaren . This was my first PR, so I thought it would be better to get your suggestions to ensure everything was correct.
An exception is thrown when an error or unexpected condition occurs during runtime, and if it’s not handled, it can cause the program to behave unpredictably. I will make sure to add the missing return and address the  suggested changes before proceeding with the PR. :)

Tejaakshaykumar committed:
Updated examples/llava/clip.cpp

added space

Co-authored-by: slaren <slarengh@gmail.com>

Tejaakshaykumar committed:
Updated clip.cpp

added return type in catch block

slaren commented on examples/llava/clip.cpp:
```diff
@@ -1270,15 +1270,24 @@ struct clip_ctx * clip_model_load(const char * fname, const int verbosity = 1) {
+            hparams.projection_dim = get_u32(ctx, format(KEY_PROJ_DIM, "vision"));
+            hparams.eps            = get_f32(ctx, format(KEY_LAYER_NORM_EPS, "vision"));
+            if (hparams.hidden_size <= 0 || hparams.n_head <= 0 || hparams.n_layer <= 0 || hparams.n_intermediate <= 0 || hparams.image_size <= 0 || hparams.patch_size <= 0 || hparams.projection_dim <= 0 || hparams.eps <= 0) {
+                LOG_TEE("Error: Invalid hyperparameter values\n");
+                return false;
+            }
+        } catch (const std::exception& e) {
+            LOG_TEE("Error while loading hyperparameters: %s\n", e.what());
+            return false;
+        }
```
This function returns a pointer, not a boolean.

slaren commented on examples/llava/clip.cpp:
```diff
@@ -1270,15 +1270,24 @@ struct clip_ctx * clip_model_load(const char * fname, const int verbosity = 1) {
+            hparams.patch_size     = get_u32(ctx, KEY_PATCH_SIZE);
+            hparams.projection_dim = get_u32(ctx, format(KEY_PROJ_DIM, "vision"));
+            hparams.eps            = get_f32(ctx, format(KEY_LAYER_NORM_EPS, "vision"));
+            if (hparams.hidden_size <= 0 || hparams.n_head <= 0 || hparams.n_layer <= 0 || hparams.n_intermediate <= 0 || hparams.image_size <= 0 || hparams.patch_size <= 0 || hparams.projection_dim <= 0 || hparams.eps <= 0) {
```
I am not sure that these checks are necessary.

slaren reviewed (changes requested):
Even ignoring the numerous errors in the very few lines of code in this PR, it doesn't even solve the linked issue since there are still many cases that can throw an exception.

@Tejaakshaykumar thank you for taking an interest on contributing to llama.cpp, but this is not the place to find a mentor. These kinds of PRs do not contribute anything of value to the project, but still waste a lot of contributor time. Please, take your time to understand the language and the code that you are working with before submitting a PR. If you are looking for somebody to mentor you, please do that in a discussion or an issue instead of a PR.
